### DISCUSSION TITLE
How or where do you call a Nuxt plugin for it to work everywhere and all the time?

### DISCUSSION THREAD

**Discussion by k16e** [nuxt/nuxt#9272] [Questions (Nuxt 2)] [answered]
Hi, I've recorded a Loom of what I'm experiencing with a plugin I wrote. Hope you don't mind taking a look https://www.loom.com/share/e3787f43e5854e038a8db7e2ee5ba885 

Specifically, what is happening is the js function in the plugin only works selectively when the DOM is fully refreshed. But works in another case all the time. And I can't figure why?

Thanks.

**Comment by clemcode** [comment, marked as answer]
Hi,

the problem you're facing is that you call `this.$toggleable()` on the `mounted()` hook of the Default **layout**. Event listeners are then correctly attached and behave as expected. 

When you navigate, your Default layout is not re-rendered, the `mounted()` hook is not called, and new event listeners are not attached to the elements on the newly created page. As your menu is part of your Default layout template, it is not destroyed and event listeners stay in place.

Only your `pages` are rendered on navigation, via the `<Nuxt />` component in your layout template. Note that Nuxt provides the `layouts` and `pages` abstractions, but this has more to do about Vue's behavior when manipulating the DOM.

---
That being said, you could call `this.$toggleable()` in the `mounted()` hook of every `page`, to make sure it runs each time a page is rendered. There are two downsides in doing so : 
- You will introduce repetition in your code.
- If you don't remove event listeners previously attached when a navigation occurs, they will pile up and risk to create a memory leak. As Vue is responsible of the DOM manipulations, it is encouraged to not manipulate it directly.

If you want a reusable Collapse component, I would go for [scoped-slots](https://vuejs.org/v2/guide/components-slots.html#Scoped-Slots), here is [an example](https://dev.to/fdietz/vue-js-headless-component-10ag) of the behavior you want to achieve

**Comment by k16e** [comment]
Thanks, @clemcode. I've just tried a POC making it a component which is always rendered/re-rendered from page to page. But, again, only the header instance works from page to page; the page instance one doesn't. I'm gonna investigate further and report on a solution. Is it a performance thing that the Default layout is not re-rendered? I'm biting to think that any js code in the Default layout shouldn't run all the time.
